lib/core/versions/latest/TransactionSelector.ts

```diff
@@ -0,0 +1,103 @@
+    if (!transactions.length) {
+      return [];
+    }
+
+    const transactionsPriorityQueue = TransactionSelector.getTransactionPriorityQueue();
+
+    const currentBlockHeight = transactions[0].transactionTime;
+    for (const transaction of transactions) {
+      if (transaction.transactionTime !== currentBlockHeight) {
+        throw new SidetreeError(ErrorCode.TransactionsNotInSameBlock, 'transaction must be in the same block to perform rate limiting, investigate and fix');
+      }
+      transactionsPriorityQueue.push(transaction);
+    }
+
+    const [numberOfOperations, numberOfTransactions] = await this.getNumberOfOperationsAndTransactionsAlreadyInBlock(currentBlockHeight);
+    let numberOfOperationsToQualify = this.maxNumberOfOperationsPerBlock - numberOfOperations;
+    let numberOfTransactionsToQualify = this.maxNumberOfTransactionsPerBlock - numberOfTransactions;
+
+    const transactionsToReturn = this.getHighestFeeTransactionsFromCurrentTransactionTime(
+      numberOfOperationsToQualify,
+      numberOfTransactionsToQualify,
+      transactionsPriorityQueue);
+
+    return transactionsToReturn;
+  }
+
+  private async getNumberOfOperationsAndTransactionsAlreadyInBlock (blockHeight: number): Promise<number[]> {
+    const transactions = await this.transactionStore.getTransactionsByTransactionTime(blockHeight);
+    let numberOfOperations = 0;
+    if (transactions) {
+      for (const transaction of transactions) {
+        const numOfOperationsInCurrentTransaction = AnchoredDataSerializer.deserialize(transaction.anchorString).numberOfOperations;
+        numberOfOperations += numOfOperationsInCurrentTransaction;
+      }
+    }
+    const numberOfTransactions = transactions ? transactions.length : 0;
+    return [numberOfOperations, numberOfTransactions];
+  }
+
+  /**
+   * Given transactions within a block, return the ones that should be processed.
+   */
+  private getHighestFeeTransactionsFromCurrentTransactionTime (
+    numberOfOperationsToQualify: number,
+    numberOfTransactionsToQualify: number,
+    transactionsPriorityQueue: any): TransactionModel[] {
+
+    let numberOfOperationsSeen = 0;
+    const transactionsToReturn = [];
+
+    while (transactionsToReturn.length < numberOfTransactionsToQualify
+      && numberOfOperationsSeen < numberOfOperationsToQualify
+      && transactionsPriorityQueue.length > 0) {
+      const currentTransaction = transactionsPriorityQueue.pop();
+      const numOfOperationsInCurrentTransaction = AnchoredDataSerializer.deserialize(currentTransaction.anchorString).numberOfOperations;
+      numberOfOperationsSeen += numOfOperationsInCurrentTransaction;
+
+      if (numberOfOperationsSeen <= numberOfOperationsToQualify) {
+        transactionsToReturn.push(currentTransaction);
+      }
+    }
+
+    // sort based on transaction number ascending
+    return transactionsToReturn.sort((a, b) => { return a.transactionNumber - b.transactionNumber; });
```

Comment:
Should we move this line back to the observer or throughput limiter, since this requirement is none of the selector's concern.

Reply:
I agree.